agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/IndyParty.kt

```diff
@@ -86,5 +87,34 @@ class IndyParty(private val webSocket: AgentWebSocketClient, val did: String, va
      * @return observable (Single<>) object emitting a single [ProofInfo] upon subscription
      */
     override fun receiveProof() = webSocket.receiveClassObject<ProofInfo>(this)
+
+    /**
+     * Returns observable ([Single]<>) object, emitting JSON-encocoded Tails file by the given tails hash.
+     *
+     * @param tailsHash string-encocoded Tails file
+     *
+     * @return observable ([Single]<>) object emitting [TailsResponse] object
+     */
+    override fun requestTails(tailsHash: String) : Single<TailsResponse> {
+        webSocket.sendClassObject(TailsRequest(tailsHash), this)
+        return webSocket.receiveClassObject(this)
+    }
+
+    private var requestHandler: (TailsRequest) -> TailsResponse = { TailsResponse(it.tailsHash, mapOf())}
+    private var tailRequestMessageHandler: (TailsRequest) -> Unit = {}
+
+    /**
+     * Sets handler for client's tails file requests
+     *
+     * @param handler a function producing [TailsResponse] from [TailsRequest]
+     */
+    override fun handleTailsRequestsWith(handler: (TailsRequest) -> TailsResponse) {
+        requestHandler = handler
+        tailRequestMessageHandler = {
```

Comment:
What will happen when I call this function twice and concurrently with request for tails?

Reply:
There's no real life use case for that, because the method is called on the side of TAILS server, once per IndyParty object. Hypothetical case when the method is called concurrently may lead to unpredictable result - when two concurrent subscribers are added on the TailsResponse message, it's not known if they are different or the same.